### ISSUE TITLE
building and running on different os with tls feature

### ISSUE THREAD

**Issue by frederikbosch** [hyperium/tonic#248]
## Bug Report
I use `rust:buster` as building system in CI because there is no official Rust Ubuntu 18.04 docker image. However, I would like to run the resulting binary on Ubuntu 18.04 (non-docker). When running the binary I get `error while loading shared libraries: libssl.so.1.1: cannot open shared object file: No such file or directory`. 

So I thought that turning on the `tls` feature would fix this as I was expecting that TLS would have been built-in into the binary then (via rustls). However, this does not seem to be the case. Am I understanding this wrong? If I am wrong, how would you approach this issue? What I want to avoid is installing a specific libssl deb in Ubuntu, but on the other hand I cannot switch operating system. Should I create a Ubuntu Rust building image myself?

### Version

```toml
tonic = { version = "0.1", features = ["tls", "transport", "codegen", "prost"] }
```

**Comment by frederikbosch** [comment]
Sorry, stupid, I should install `libssl-dev` on my Ubuntu machine.

**Comment by LucioFranco** [comment]
I don't think we require `libssl-dev` maybe there is another library pulling it in somehow? Last I checked rustls doesn't require it?

**Comment by frederikbosch** [comment]
Hmm, I think you are right. I missed that. I think I also have to activate rustls in Rusoto (Rust S3 lib). Thanks for the suggestion.

**Comment by frederikbosch** [comment]
Minor question @LucioFranco: why do you (and libraries like Rusoto and reqwest) do not use rustls by default? It seems much easier for the user of those libraries. But since everybody is doing it, there must be a reason for it.

**Comment by LucioFranco** [comment]
I think by default they will use the default system ssl, which in your case on linux is openssl, while rustls basically works without deps. The reason I chose rustls is for two main reasons 1) native-tls doesn't currently support ALPN negotiation which is important for http2 2) again native-tls is missing many features needed for things like mTLS. I then chose to support rustls over openssl because its prob safer, quicker to build and closest to pure rust. The goal of tonic to have a pure rust gRPC lib so it fits that very well. At somepoint soon ill write a `tonic-openssl` lib.
